internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -0,0 +1,504 @@
+)
+
+// reconcileResult is an interface that encapsulates the result of a Reconcile
+// call, as returned by the action corresponding to the current state.
+type reconcileResult interface {
+	Result() (reconcile.Result, error)
+	Dirty() bool
+}
+
+// reconcileComplete is a result indicating that the current reconcile has completed,
+// and there is nothing else to do.
+type reconcileComplete struct {
+	dirty bool
+}
+
+func (r reconcileComplete) Result() (result reconcile.Result, err error) {
+	return
+}
+
+func (r reconcileComplete) Dirty() bool {
+	return r.dirty
+}
+
+// reconcileRequeue is a result indicating that the reconcile should be
+// requeued.
+type reconcileRequeue struct {
+	delay time.Duration
+}
+
+func (r reconcileRequeue) Result() (result reconcile.Result, err error) {
+	result.RequeueAfter = r.delay
+	// Set Requeue true as well as RequeueAfter in case the delay is 0.
+	result.Requeue = true
+	return
+}
+
+func (r reconcileRequeue) Dirty() bool {
+	return false
+}
+
+// reconcileError is a result indicating that an error occurred while attempting
+// to advance the current reconcile, and that reconciliation should be retried.
+type reconcileError struct {
+	err error
+}
+
+func (r reconcileError) Result() (result reconcile.Result, err error) {
+	err = r.err
+	return
+}
+
+func (r reconcileError) Dirty() bool {
+	return false
+}
+
+// +kubebuilder:rbac:groups=metal3.io,resources=baremetalhosts,verbs=get;list;watch;update;patch
+
+func (r *BMACReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	bmh := &bmh_v1alpha1.BareMetalHost{}
+
+	if err := r.Get(ctx, req.NamespacedName, bmh); err != nil {
+		if !errors.IsNotFound(err) {
+			return reconcileError{err}.Result()
+		}
+		return reconcileComplete{}.Result()
+	}
+
+	// Let's reconcile the BMH
+	result := r.reconcileBMH(ctx, bmh)
+	if result.Dirty() {
+		err := r.Client.Update(ctx, bmh)
+		if err != nil {
+			r.Log.WithError(err).Errorf("Error updating hardwaredetails")
+			return reconcileError{err}.Result()
+		}
+
+		return result.Result()
+	}
+
+	// Should we check the status of the BMH resource?
+	// Provisioning / Deprovisioning ?
+	// What happens if we do the reconcile on a BMH that
+	// is in a Deprovisioning state?
+
+	agent := r.findAgent(ctx, bmh)
+
+	// handle multiple agents matching the
+	// same BMH's Mac Address
+
+	if agent == nil {
+		return result.Result()
+	}
+
+	result = r.reconcileAgentInventory(bmh, agent)
+	if result.Dirty() {
+		err := r.Client.Update(ctx, bmh)
+		if err != nil {
+			r.Log.WithError(err).Errorf("Error updating hardwaredetails")
+			return reconcileError{err}.Result()
+		}
+	}
+
+	// if we get to this point, it means we have found an Agent that is associated
+	// with the BMH being reconciled. We will call both, reconcileAgentSpec and
+	// reconcileAgentInventory, every time. The logic to decide whether there's
+	// any action to take is implemented in each function respectively.
```

Comment:
This comment makes more sense before the call to reconcileAgentInventory()

Reply:
Lol, yeah! I moved the call and then didn't move the comment :D